### PR TITLE
added Parakeet benchmarks to arc_distance and pairwise, updated cold/warm timings

### PR DESCRIPTION
**Pull request by iskandr**
Hey Olivier, 

Sorry if this is too much stuff in one pull request, I can split things up if you'd like. 

There are a few changes here:
- parallel Parakeet implementation of arc distance 
- parallel Parakeet implementation of pairwise distances
- run_benchmarks.py: only disables cold/warm timings for pure Python functions

### CODE REVIEW

**Comment by ogrisel** [comment]
Looks good but github complains about the merge. Can you please rebase on top of the current master?

**Comment by iskandr** [comment]
Honestly, Olivier, I have no idea what just happened at the commandline. I need to watch a lecture on git, I still think of it as "local svn and then magic".

**Comment by ogrisel** [comment]
Haha :) Could you please squash all the commits into one if are willing to learn more git wizardry? Otherwise I will do it myself.

File: arc_distance/arc_distance_parakeet.py [hunk 1]
```diff
@@ -3,8 +3,28 @@
...
+  def arc_dist(ai, bj):
+    theta1 = ai[0]
+    phi1 = ai[1]
+    theta2 = bj[0]
+    phi2 = bj[1]
+    d_theta = theta2 - theta1
+    d_phi = phi2 - phi1
+    temp = (np.sin(d_theta / 2) ** 2) + (np.cos(theta1) * np.cos(theta2) * np.sin(d_phi / 2) ** 2)
+    return 2 * np.arctan2(np.sqrt(temp), np.sqrt(1 - temp))
+  return np.array([[arc_dist(ai,bj) for bj in b] for ai in a])
```
Comment (ogrisel) [inline arc_distance/arc_distance_parakeet.py]: pep8: `arc_dist(ai, bj)`

File: arc_distance/arc_distance_parakeet.py [hunk 2]
```diff
@@ -3,8 +3,28 @@
...
+  by Parakeet. 
+  """
+  def arc_dist(ai, bj):
+    theta1 = ai[0]
+    phi1 = ai[1]
+    theta2 = bj[0]
+    phi2 = bj[1]
+    d_theta = theta2 - theta1
+    d_phi = phi2 - phi1
+    temp = (np.sin(d_theta / 2) ** 2) + (np.cos(theta1) * np.cos(theta2) * np.sin(d_phi / 2) ** 2)
```
Comment (ogrisel) [inline arc_distance/arc_distance_parakeet.py]: pep8:

```
temp = (np.sin(d_theta / 2) ** 2)
        + (np.cos(theta1) * np.cos(theta2) * np.sin(d_phi / 2) ** 2)
```

**Comment by iskandr** [comment]
After some misadventures with rebasing, I think I need external assistance. I tried git rebase -i HEAD~5 but I just end up having to re-merge old conflicts.

**Comment by ogrisel** [comment]
You can `git rebase --abort` to abort the interactive rebase then squash with a simple reset: `git reset HEAD~5` then `git add .` and finally `git commit`.

**Comment by ogrisel** [comment]
`rebase -i` is only useful it you want to fine tune your squash (for instance by changing the order of independent commit before squashing a subset of them or editing the individual commit messages).

**Comment by iskandr** [comment]
What's the easy/intelligent way to handle the fallout of git rebase HEAD~5?

First, rewinding head to replay your work on top of it...
Applying: added Parakeet benchmarks to arc_distance and pairwise, updated
cold/warm timings
Using index info to reconstruct a base tree...
<stdin>:15: trailing whitespace.
import numpy as np
<stdin>:17: trailing whitespace.
@jit
<stdin>:22: trailing whitespace.
  by Parakeet.
<stdin>:29: trailing whitespace.
    d_theta = theta2 - theta1
<stdin>:48: trailing whitespace.
import numpy as np
warning: squelched 7 whitespace errors
warning: 12 lines add whitespace errors.
Falling back to patching base and 3-way merge...
Auto-merging run_benchmarks.py
CONFLICT (content): Merge conflict in run_benchmarks.py
Failed to merge in the changes.
Patch failed at 0001 added Parakeet benchmarks to arc_distance and
pairwise, updated cold/warm timings

When you have resolved this problem run "git rebase --continue".
If you would prefer to skip this patch, instead run "git rebase --skip".
To check out the original branch and stop rebasing run "git rebase --abort".

**Comment by iskandr** [comment]
I think that worked, but can you make sure I didn't do anything terribly stupid? 

Also, it seems that being aware of the sequential application of diffs seems pretty important for unraveling git's mysteries.

**Comment by iskandr** [comment]
Thanks!

It looks like Numba now raises this error in all of its benchmarks:
"AttributeError:
'NoneType' object has no attribute 'core'"
...maybe it's a mismatch between the Numba and llvmpy versions?

Also, we should probably add more varied benchmarks to spice things up.
I'll try writing a few different styles of decision tree threholding.

**Comment by ogrisel** [comment]
Yes I saw about the numba pbm but only once I updated. I think an operating system update broke my llvm install on my laptop. I should try to use anaconda or mini-conda to re-install numba but I was too lazy. If you can run the benchmarks successfully on you box (check the generated report folder), you could push the github.io website with `make github` after a `pip install ghp-import`.

**Comment by ogrisel** [comment]
You can ask @aterrel for the push rights if you don't have them already. Apparently I cannot grant them to you myself.

**Comment by aterrel** [comment]
Added @iskandr . Hopefully, I'll have more time after the holidays to help out.

**Comment by iskandr** [comment]
Awesome, thanks guys. I'm busy with my thesis right now, but when I get
some time I'll try to spruce up the benchmark selection and re-run numbers
with a working Numba.

**Pull request merged**